### PR TITLE
Enforce a decoding limit in MultiAssets

### PR DESCRIPTION
**Pull request by franciscoaguirre**
# Description

Added a limit to the number of assets that can be decoded when decoding a `MultiAssets` instance

### CODE REVIEW

**Comment by franciscoaguirre** [comment]
bot fmt

File: polkadot/xcm/src/v3/multiasset.rs [hunk 1]
```diff
@@ -517,9 +516,16 @@ impl MaxEncodedLen for MultiAssets {
 }
 
 impl Decode for MultiAssets {
-	fn decode<I: codec::Input>(input: &mut I) -> Result<Self, parity_scale_codec::Error> {
-		Self::from_sorted_and_deduplicated(Vec::<MultiAsset>::decode(input)?)
-			.map_err(|()| "Out of order".into())
+	fn decode<I: codec::Input>(input: &mut I) -> Result<Self, codec::Error> {
```
Comment (ggwpez) [inline polkadot/xcm/src/v3/multiasset.rs:519]: Without knowing the context here; I suppose we dont need to track this globally via environmental?  
I hope its enough just in combination with the overall instruction + nesting limit, but its good to double-check.

Comment (franciscoaguirre) [reply polkadot/xcm/src/v3/multiasset.rs:519]: There's no way to nest `MultiAsset`s, so it's okay to not track anything globally.
Each individual `MultiAssets` being decoded will be able to have at most `MAX_ITEMS_IN_MULTIASSETS = 20` assets.
The limit set on instructions will make sure you don't add a lot of these.

File: polkadot/xcm/src/v3/multiasset.rs [hunk 2]
```diff
@@ -974,4 +980,31 @@ mod tests {
...
+
+		// Having lots of one asset will work since they are deduplicated
+		let lots_of_one_asset: MultiAssets =
+			vec![(GeneralIndex(1), 1u128).into(); MAX_ITEMS_IN_MULTIASSETS + 1].into();
+		let encoded = lots_of_one_asset.encode();
+		assert!(MultiAssets::decode(&mut &encoded[..]).is_ok());
+
+		// Fewer assets than the limit works
+		let mut few_assets: MultiAssets = Vec::new().into();
+		for i in 0..MAX_ITEMS_IN_MULTIASSETS - 1 {
```
Comment (KiChjang) [inline polkadot/xcm/src/v3/multiasset.rs:996]: We want to test the edge case as much as possible, so let's do this:

```suggestion
		for i in 0..MAX_ITEMS_IN_MULTIASSETS {
```

**Review by KiChjang — APPROVED** [review]
Looks good after resolving my comment.

**Comment by franciscoaguirre** [comment]
@ggwpez Can you check again please?

File: polkadot/xcm/src/v3/multiasset.rs [hunk 3]
```diff
@@ -506,8 +506,7 @@ impl TryFrom<OldMultiAsset> for MultiAsset {
 #[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
 pub struct MultiAssets(Vec<MultiAsset>);
 
-/// Maximum number of items we expect in a single `MultiAssets` value. Note this is not (yet)
-/// enforced, and just serves to provide a sensible `max_encoded_len` for `MultiAssets`.
+/// Maximum number of items in a single `MultiAssets` value that can be decoded.
 const MAX_ITEMS_IN_MULTIASSETS: usize = 20;
```
Comment (ggwpez) [inline polkadot/xcm/src/v3/multiasset.rs:510]: I would always make this stuff public, otherwise downstream crates will be confused why an opaque number is used.

File: polkadot/xcm/src/v3/multiasset.rs [hunk 4]
```diff
@@ -517,9 +516,16 @@ impl MaxEncodedLen for MultiAssets {
...
+	fn decode<I: codec::Input>(input: &mut I) -> Result<Self, codec::Error> {
+		let number_of_assets: u32 = <codec::Compact<u32>>::decode(input)?.into();
+		if number_of_assets > MAX_ITEMS_IN_MULTIASSETS as u32 {
+			return Err(codec::Error::from("Max MultiAssets exceeded"))
+		}
+		Self::from_sorted_and_deduplicated(codec::decode_vec_with_len(
+			input,
+			number_of_assets as usize,
+		)?)
+		.map_err(|()| "Out of order".into())
```
Comment (ggwpez) [inline polkadot/xcm/src/v3/multiasset.rs:528]: You can use the `BoundedVec::<..>::decode(..)?.into_inner()` here again, or? Anyway it should both work.

**Review by ggwpez — APPROVED** [review]

File: polkadot/xcm/src/v3/multiasset.rs [hunk 5]
```diff
@@ -517,9 +516,16 @@ impl MaxEncodedLen for MultiAssets {
 }
 
 impl Decode for MultiAssets {
-	fn decode<I: codec::Input>(input: &mut I) -> Result<Self, parity_scale_codec::Error> {
-		Self::from_sorted_and_deduplicated(Vec::<MultiAsset>::decode(input)?)
-			.map_err(|()| "Out of order".into())
+	fn decode<I: codec::Input>(input: &mut I) -> Result<Self, codec::Error> {
+		let number_of_assets: u32 = <codec::Compact<u32>>::decode(input)?.into();
+		if number_of_assets > MAX_ITEMS_IN_MULTIASSETS as u32 {
+			return Err(codec::Error::from("Max MultiAssets exceeded"))
```
Comment (bkchr) [inline polkadot/xcm/src/v3/multiasset.rs:522]:
```suggestion
			return Err("Max MultiAssets exceeded".into())
```

File: polkadot/xcm/src/v3/multiasset.rs [hunk 6]
```diff
@@ -517,9 +516,16 @@ impl MaxEncodedLen for MultiAssets {
 }
 
 impl Decode for MultiAssets {
-	fn decode<I: codec::Input>(input: &mut I) -> Result<Self, parity_scale_codec::Error> {
-		Self::from_sorted_and_deduplicated(Vec::<MultiAsset>::decode(input)?)
-			.map_err(|()| "Out of order".into())
+	fn decode<I: codec::Input>(input: &mut I) -> Result<Self, codec::Error> {
+		let number_of_assets: u32 = <codec::Compact<u32>>::decode(input)?.into();
```
Comment (bkchr) [inline polkadot/xcm/src/v3/multiasset.rs:520]:
```suggestion
		let number_of_assets = <codec::Compact<u32>>::decode(input)?.0;
```

**Review by bkchr — APPROVED** [review]
